### PR TITLE
#162170709 Validate incoming signup/registration data

### PR DESCRIPTION
**Pull request by luqmanoop**
#### What does this PR do?
Validates incoming user requests on signup/registration route with descriptive error messages

*Other actions*
- Write tests for validated data
- Write test authentication for middleware

#### Description of Task to be completed?
Validate incoming data in `/api/users/signup` route

#### How should this be manually tested?
- Clone the repo and `cd` into the project
- Run `npm run dev:start` or `npm run dev:start:watch`
- Send a request to http://localhost:3000/api/users/signup with different (valid/invalid) data

##### Run the test
`npm run test` or `npm run test:watch`

#### Any background context you want to provide?
This features makes sure users don't signup on our site with invalid data

#### What are the relevant pivotal tracker stories?
[#162170709](https://www.pivotaltracker.com/story/show/162170709)

#### Screenshots (if appropriate)
*Validations*
screen shot 2018-12-04 at 10 56 49 am

*Test coverage*
screen shot 2018-12-05 at 10 58 55 am

#### Questions:
N/A

### CODE REVIEW

**Review by Johnsonojo — CHANGES_REQUESTED** [review]
Hi codeshifu, nice job on validations and testing. While reviewing your code, I noticed some inconsistencies in your tests. You used `req`, `res`, `request`, and `response` interchangeably. I believe you can refactor and make use of `request` and `response` for consistency.

**Review by Johnsonojo — APPROVED** [review]

File: src/controllers/users.js [hunk 1]
```diff
@@ -34,54 +32,61 @@ class userHandler {
         id: userInfo.id,
         username: userInfo.username
       };
+
       const time = {};
       time.expiresIn = '24h';
-      const token = generateToken(payload, time);
-      response.status(201)
-        .json({
+      try {
```
Comment (oluwajuwon) [inline src/controllers/users.js:38]: I like what you did here, nice observation

**Review by marcdomain — APPROVED** [review]
This looks good to me! Nice job @codeshifu

**Review by oluwajuwon — APPROVED** [review]
Nice job @codeshifu This looks good to me👍🏿

**Comment by omokaro-faith** [comment]
@codeshifu resolve conflicts.

File: package.json [hunk 2]
```diff
@@ -7,12 +7,14 @@
...
     "dev": "npm-run-all --parallel clean build migrate",
     "dev:start": "npm run dev && npm run start",
+    "dev:start:watch": "nodemon --watch src --exec 'npm run dev:start'",
     "start": "node dist/server.js",
     "build": "babel ./src -d dist",
     "migrate": "sequelize db:migrate",
     "migrate:test": "sequelize db:migrate:undo:all && npm run migrate",
     "postinstall": "npm run dev",
     "test": "export NODE_ENV=test && npm run migrate:test && nyc mocha --recursive --require babel-register --exit",
+    "test:watch": "nodemon --watch test --exec npm test",
```
Comment (akhilome) [inline package.json:17]: Thanks, a ton, for adding this (and `dev:start:watch`) in! 🙏

**Review by akhilome — APPROVED** [review]
LGTM! 👍

File: package.json (hunk 2 above)
Comment (luqmanoop) [reply package.json:17]: Thanks for taking the time to review. Much appreciated 👍

**Pull request merged**